### PR TITLE
Spline Fit for Recorder.plot() and AccumulateStepper(LearnerCallback)

### PR DESCRIPTION
**Pull request by KeremTurgutlu**
**Recorder.plot()** 

Here is the discussion thread for this PR:

https://forums.fast.ai/t/autolrfinder/38331/22

So far this method seems to be working pretty well for fixing outlier problems with lr vs loss plot.
Although, probably it's best to see if there is enough need for this feature, so please thumbs up if you see a fit here. It doesn't have drastic changes in `.plot()` method code. Now it optionally takes argument `k` as the spline order. If no k if given original smoothen loss is plotted.

**AccumulateStepper(LearnerCallback)**

Here is the discussion thread for this PR:

https://forums.fast.ai/t/accumulating-gradients/33219/26

Further experiments can be made...

### CODE REVIEW

File: fastai/basic_train.py [hunk 1]
```diff
@@ -469,18 +469,27 @@ def plot_lr(self, show_moms=False)->None:
...
+        losses = [x.item() for x in losses]
+        if 'k' in kwargs: losses = self.smoothen_by_spline(lrs, losses, **kwargs)
         _, ax = plt.subplots(1,1)
         ax.plot(lrs, losses)
         ax.set_ylabel("Loss")
         ax.set_xlabel("Learning Rate")
         ax.set_xscale('log')
         ax.xaxis.set_major_formatter(plt.FormatStrFormatter('%.0e'))
-        mg = (np.gradient(np.array([x.item() for x in losses]))).argmin()
+        mg = (np.gradient(np.array(losses))).argmin()
```
Comment (sgugger) [inline fastai/basic_train.py]: I think you need the `[x.item() for x in losses]` since the element of `losses` are rank-0 tensors.

File: fastai/train.py [hunk 2]
```diff
@@ -94,8 +94,52 @@ def clip_grad(learn:Learner, clip:float=0.1)->Learner:
     "Add gradient clipping of `clip` during training."
     learn.callback_fns.append(partial(GradientClipping, clip=clip))
     return learn
-
 Learner.clip_grad = clip_grad
+     
+class AccumulateStepper(LearnerCallback):
+    """
```
Comment (sgugger) [inline fastai/train.py]: On one line with just " " (fastai style guide).

File: fastai/train.py [hunk 3]
```diff
@@ -94,8 +94,52 @@ def clip_grad(learn:Learner, clip:float=0.1)->Learner:
...
+    Does accumlated step every nth step by accumulating gradients
+    """
+    def __init__(self, learn:Learner, n_step:int = 1):
+        super().__init__(learn)
+        self.n_step = n_step
+ 
+    def on_train_begin(self, **kwargs):
+        "check if loss is reduction"
+        if hasattr(self.loss_func, "reduction") and (self.loss_func.reduction == "mean"):
+             print("For better gradients consider 'reduction=sum'")
```
Comment (sgugger) [inline fastai/train.py]: Make that a warning instead.

File: fastai/train.py [hunk 4]
```diff
@@ -94,8 +94,52 @@ def clip_grad(learn:Learner, clip:float=0.1)->Learner:
...
+        
+    def on_epoch_begin(self, **kwargs):
+        "init samples and batches, change optimizer"
+        self.acc_samples = 0
```
Comment (sgugger) [inline fastai/train.py]: The two init on one line bli,bla = 0,0 (fastai style guide).

File: fastai/train.py [hunk 5]
```diff
@@ -94,8 +94,52 @@ def clip_grad(learn:Learner, clip:float=0.1)->Learner:
...
+        if (self.acc_batches % self.n_step) == 0:
+            for p in (self.learn.model.parameters()):
+                if p.requires_grad: p.grad.div_(self.acc_samples)
+            self.acc_samples = 0
+            return False
+        else: return True
+    
+    def on_step_end(self, **kwargs):
+        "zero gradients after stepping"
+        if (self.acc_batches % self.n_step) == 0: return False
```
Comment (sgugger) [inline fastai/train.py]: Just returning the result of the test is enough ;)

File: fastai/train.py [hunk 6]
```diff
@@ -94,8 +94,52 @@ def clip_grad(learn:Learner, clip:float=0.1)->Learner:
...
+        else: return True
+    
+    def on_epoch_end(self, **kwargs):
+        "step the rest of the accumulated grads if not perfectly divisible"
+        for p in (self.learn.model.parameters()):
+                if p.requires_grad: p.grad.div_(self.acc_samples)
+        self.learn.opt.step()
```
Comment (sgugger) [inline fastai/train.py]: Maybe add a `drop_last` argument similar to the one in the pytorch `DataLoader`? If it's `True`, we won't do the step, just zeroing.

**Comment by sgugger** [comment]
Thanks a lot! There are a few style change to make and I made a suggestion to add an argument, then we'll be able to merge this.

**Comment by KeremTurgutlu** [comment]
Thanks for suggestions :) I will update the PR as soon as possible, probably tonight.

File: fastai/train.py (hunk 5 above)
Comment (KeremTurgutlu) [reply fastai/train.py]: I didn't understand this part, can you elaborate please :)

File: fastai/basic_train.py [hunk 7]
```diff
@@ -469,18 +469,27 @@ def plot_lr(self, show_moms=False)->None:
...
+        xs = np.arange(len(ys))
+        spl = scipy.interpolate.UnivariateSpline(xs, ys, **kwargs)
+        ys = spl(xs)
+        return ys
+
+    def plot(self, skip_start:int=10, skip_end:int=5, **kwargs)->None:
         "Plot learning rate and losses, trimmed between `skip_start` and `skip_end`. Optionally plot and return min gradient"
         lrs = self.lrs[skip_start:-skip_end] if skip_end > 0 else self.lrs[skip_start:]
         losses = self.losses[skip_start:-skip_end] if skip_end > 0 else self.losses[skip_start:]
+        losses = [x.item() for x in losses]
```
Comment (KeremTurgutlu) [inline fastai/basic_train.py]: That's true that's why I am getting item here.

Comment (sgugger) [reply fastai/basic_train.py]: Eh, how did I miss this line?

File: fastai/train.py (hunk 5 above)
Comment (sgugger) [reply fastai/train.py]: No I was saying, `return (self.acc_batches % self.n_step) != 0` (we need `False` when we should do the step + zero, `True` if we skip)

Instead of testing if bool return False else return True, just return not book ;)

File: fastai/train.py [hunk 8]
```diff
@@ -94,8 +94,49 @@ def clip_grad(learn:Learner, clip:float=0.1)->Learner:
...
     return learn
-
 Learner.clip_grad = clip_grad
+     
+class AccumulateStepper(LearnerCallback):
+    "Does accumlated step every nth step by accumulating gradients"
+
+    def __init__(self, learn:Learner, n_step:int = 1, drop_last:bool = False):
+        super().__init__(learn)
+        self.n_step = n_step
```
Comment (sgugger) [inline fastai/train.py]: Both on one line (fastai style guide)

**Comment by sgugger** [comment]
Two changes (one cosmetic, one more important in on_step_end) and we'll be good to go, thanks again!

**Comment by sgugger** [comment]
Thanks! Now we'll have to address that issue in BatchNorm.
